Write your own VM: Builtins: Execution logic (auto-deductions) (#62)
lambdaclass/cairo-vm.go: README.md

}
```

Now we will dive deeper into how `auto-deduction` rules come into play during execution:

Before builtins, the basic flow for computing the value of an operand was to first compute its address, and then if we couldn't find it in memory, we would deduce its value based on the other operands.
With the introduction of builtins and their auto-deduction rules, this flow changes a bit. Now we compute the address, use it to fetch the value from memory, if we can't find it in memory we try to use the builtin's auto deduction rules, and if we can't deduce it via builtins we will then deduce it based on the other operands's.
But what does it mean to use the builtin's auto deduction rules to deduce the value of an operand?

##### DeduceMemoryCell

> This method will iterate over the builtin runners and try to find a builtin who's base's segment index matches the operand to be deduced's address. That is to say, it checks if the address' belongs to a builtin's segment. If a match is found, it uses the builtin's `DeduceMemoryCell` method to run the builtin's auto-deduction rules and calculare the value of the operand

Reviewer: ```suggestion
This method will iterate over the builtin runners and try to find a builtin who's base's segment index matches the operand to be deduced's address. That is to say, it checks if the address belongs to a builtin's segment. If a match is found, it uses the builtin's `DeduceMemoryCell` method to run the builtin's auto-deduction rules and calculate the value of the operand
```

> This method will iterate over the builtin runners and try to find a builtin who's base's segment index matches the operand to be deduced's address. That is to say, it checks if the address' belongs to a builtin's segment. If a match is found, it uses the builtin's `DeduceMemoryCell` method to run the builtin's auto-deduction rules and calculate the value of the operand

Reviewer: ```suggestion
This method will iterate over the builtin runners and try to find a builtin who's base's segment index matches the operand to be deduced's address. That is to say, it checks if the address belongs to a builtin's segment. If a match is found, it uses the builtin's `DeduceMemoryCell` method to run the builtin's auto-deduction rules and calculate the value of the operand
```

  if err != nil {
   return *memory.NewMaybeRelocatableFelt(lambdaworks.FeltZero()), err
  }
  if res == nil {
   res = deducedRes
  }
 }
 if op1 != nil {
  vm.Segments.Memory.Insert(op1_addr, op1)
 } else {
  return *memory.NewMaybeRelocatableFelt(lambdaworks.FeltZero()), errors.New("Failed to compute or deduce op1")
 }
 return *op1, nil
}
```

Now we integrate these two new methods into our previous `ComputeOperands` method:

```go
func (vm *VirtualMachine) ComputeOperands(instruction Instruction) (Operands, error) {
 var res *memory.MaybeRelocatable

 dst_addr, err := vm.RunContext.ComputeDstAddr(instruction)
 if err != nil {
  return Operands{}, errors.New("FailedToComputeDstAddr")
 }
 dst, _ := vm.Segments.Memory.Get(dst_addr)

 op0_addr, err := vm.RunContext.ComputeOp0Addr(instruction)
 if err != nil {
  return Operands{}, fmt.Errorf("FailedToComputeOp0Addr: %s", err)
 }
 op0_op, _ := vm.Segments.Memory.Get(op0_addr)

 op1_addr, err := vm.RunContext.ComputeOp1Addr(instruction, op0_op)
 if err != nil {
  return Operands{}, fmt.Errorf("FailedToComputeOp1Addr: %s", err)
 }
 op1_op, _ := vm.Segments.Memory.Get(op1_addr)

>     var op0 memory.MaybeRelocatable

Reviewer 2: ```suggestion
  var op0 memory.MaybeRelocatable
```

> 	op1_op, _ := vm.Segments.Memory.Get(op1_addr)
> 
>     var op0 memory.MaybeRelocatable
> 	if op0_op != nil {

Reviewer: ```suggestion
	op1_op, _ := vm.Segments.Memory.Get(op1_addr)

        var op0 memory.MaybeRelocatable
	if op0_op != nil {
```

		op0 = *op0_op
	} else {
		op0, res, err = vm.ComputeOp0Deductions(op0_addr, &instruction, dst, op1_op)
		if err != nil {
			return Operands{}, err
		}
	}

	var op1 memory.MaybeRelocatable
	if op1_op != nil {
		op1 = *op1_op
	} else {
		op1, err = vm.ComputeOp1Deductions(op1_addr, &instruction, dst, op0_op, res)
		if err != nil {
			return Operands{}, err
		}

> 	}
>     if res == nil {
> 		res, err = vm.ComputeRes(instruction, op0, op1)

Reviewer: ```suggestion
	}
        if res == nil {
		res, err = vm.ComputeRes(instruction, op0, op1)
```
